sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesSplittableDoFnTest.java

```diff
@@ -0,0 +1,520 @@
+import org.apache.beam.sdk.values.PCollection;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Tests for {@link DoFnSignatures} focused on methods related to <a
+ * href="https://s.apache.org/splittable-do-fn">splittable</a> {@link DoFn}.
+ */
+@SuppressWarnings("unused")
+@RunWith(JUnit4.class)
+public class DoFnSignaturesSplittableDoFnTest {
+  @Rule public ExpectedException thrown = ExpectedException.none();
+
+  private static class SomeRestriction {}
+
+  private abstract static class SomeRestrictionTracker
+      implements RestrictionTracker<SomeRestriction> {}
+
+  private abstract static class SomeRestrictionCoder implements Coder<SomeRestriction> {}
+
+  @Test
+  public void testReturnsProcessContinuation() throws Exception {
+    DoFnSignature.ProcessElementMethod signature =
+        analyzeProcessElementMethod(
+            new AnonymousMethod() {
+              private DoFn.ProcessContinuation method(
+                  DoFn<Integer, String>.ProcessContext context) {
+                return null;
+              }
+            });
+
+    assertTrue(signature.hasReturnValue());
+  }
+
+  @Test
+  public void testHasRestrictionTracker() throws Exception {
+    DoFnSignature.ProcessElementMethod signature =
+        analyzeProcessElementMethod(
+            new AnonymousMethod() {
+              private void method(
+                  DoFn<Integer, String>.ProcessContext context, SomeRestrictionTracker tracker) {}
+            });
+
+    assertTrue(signature.isSplittable());
+    assertTrue(signature.extraParameters().contains(DoFnSignature.Parameter.RESTRICTION_TRACKER));
+    assertEquals(SomeRestrictionTracker.class, signature.trackerT().getRawType());
+  }
+
+  @Test
+  public void testSplittableProcessElementMustNotHaveOtherParams() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("must not have any extra context arguments");
+    thrown.expectMessage("BOUNDED_WINDOW");
+
+    DoFnSignature.ProcessElementMethod signature =
+        analyzeProcessElementMethod(
+            new AnonymousMethod() {
+              private void method(
+                  DoFn<Integer, String>.ProcessContext context,
+                  SomeRestrictionTracker tracker,
+                  BoundedWindow window) {}
+            });
+  }
+
+  @Test
+  public void testInfersBoundedness() throws Exception {
+    class BaseFn extends DoFn<Integer, String> {
+      @ProcessElement
+      public void processElement(ProcessContext context, SomeRestrictionTracker tracker) {}
+
+      @GetInitialRestriction
+      public SomeRestriction getInitialRestriction(Integer element) {
+        return null;
+      }
+
+      @NewTracker
+      public SomeRestrictionTracker newTracker(SomeRestriction restriction) {
+        return null;
+      }
+    }
+
+    @Bounded
+    class BoundedFn extends BaseFn {}
```

Comment:
Local classes tend to be iffy, but I guess this is OK. Would still be good to split this test up into InfersUnboundedness and InfersBoundedness.